Java异常(待学习)
2018/3/5: 
异常分为两个大类：Error类和Exception类；Error类代表了编译核系统错误；Exception类包含着运行异常Runtime_Exception 和非运行异常Non_RuntimeException（可检测异常）：java.util,java.lang,java.net均为非运行异常。

compiler 不对运行异常进行检查（任何情况都可能发生，为了避免巨大系统开销，故不对运行异常进行检查），运行异常对应编译错误；运行错误就表面代码有错误比如数组越界 or除以0 等都是错误。

对于可能可检测异常，方法或者构造方法的throw的子句必须列出该异常对于的类名


Java异常通常会有一下几种情况：(1)用户输入了非法数据;(2)要打开的文件不存在;(3)网络通信时连接中断，或者JVM内存溢出 这三种错误是由于用户错误引起，还有其他一些是由于物理错误引起。

Java异常处理机制：
(1)检查性异常：最具代表的检查性异常是用户错误或问题引起的异常，这是程序员无法预见的。如打开一个已被删除的文件路径；这些异常在编译时不能被简单忽略
(2)运行时异常：运行时异常是可能被程序员避免的异常，可在运行时被忽略
(3)错误：错误不是异常，而是脱离程序员控制的问题。错误在代码中被忽略：如堆溢出，错误发生，在编译时，检查不到

所有异常都是java.lang.Exception类继承的子类

异常是一些错误；但错误不是异常

Throwable下面有Exception类和Error类；Exception类下面有IOException类和RuntimeException类

Java 程序通常不捕获错误。错误一般发生在严重故障时，它们在Java程序处理的范畴之外。

Error用来指示运行时环境发生的错误。如JVM从内存中溢出，程序不会错误中恢复。


Java内置异常类：非检查性，检查性异常

异常方法：Throwable类主要方法
(1)public String getMessage（）
(2)public Throwable getCause()
(3)public String toString()
(4)public void printStackTrace()
(5)public StackTraceElement[] getStackTrace()
(6)public Throwable fillInStackTrace()


捕获异常
try/catch代码块中代码被称为保护代码块
try
{
  //代码
}catch(Exception name,e1)
{
  catch 块
}



多重块捕获
try
{
  //代码块
}catch(异常类型1，异常变量名1)
{
  //代码块
}catch(异常类型2，异常变量名2)
{
  //代码块
}catch(异常类型2，异常变量名2)
{
  //代码块
}

throws/throw关键字：若一个方法未捕获一个检查性异常，该方法必须用throws来声明，throws放在方法签名尾部；亦可用throw抛出一个异常不论其为新实例化还是刚捕捉到的。

finally 关键字：用于创建try后面执行代码；无论异常是否发生，finally代码块中代码都会执行；finally代码块一般位于catch代码最后，一般是一些善后的语句






